### PR TITLE
feat: add WebFrame.windowFeatures property

### PR DESCRIPTION
**Pull request by samuelmaddock**
#### Description of Change

Adds a `windowFeatures` property to the renderer's WebFrame API. This exposes the raw string of features passed into `window.open(url, name, features)`.

Electron has a [Chromium patch](https://github.com/electron/electron/blob/954fd725009093660599a03a6926d57717f5a8c4/patches/chromium/can_create_window.patch) which adds the raw feature list string. I'd like to be able to access it from a preload script without needing to hook up an IPC to save and retrieve this data already present in the renderer process.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added WebFrame.windowFeatures API to access features list passed to `window.open(url, name, features)`.

### CODE REVIEW

File: docs/api/web-frame.md [hunk 1]
```diff
@@ -314,3 +314,8 @@ current renderer process.
 An `Integer` representing the unique frame id in the current renderer process.
 Distinct WebFrame instances that refer to the same underlying frame will have
 the same `routingId`.
+
+### `webFrame.windowFeatures` _Readonly_
+
+A `String` representing the encoded list of window features passed into
+`window.open(url, target, windowFeatures)`.
```
Comment (nornagon) [inline docs/api/web-frame.md:321]: This is much too concise! Please expand this to make it clearer when this is present, what happens upon navigation, etc.

File: shell/renderer/api/electron_api_web_frame.cc [hunk 2]
```diff
@@ -904,6 +907,16 @@ class WebFrameRenderer : public gin::Wrappable<WebFrameRenderer>,
...
   }
+
+  std::string GetWindowFeatures(v8::Isolate* isolate) {
+    content::RenderFrame* render_frame;
+    if (!MaybeGetRenderFrame(isolate, "windowFeatures", &render_frame))
+      return "";
+
+    blink::LocalFrameToken frame_token = render_frame->GetWebFrame()->GetLocalFrameToken();
+    blink::LocalFrame* frame = blink::LocalFrame::FromFrameToken(frame_token);
+    return frame ? frame->GetPage()->GetWindowFeatures().raw_features.Utf8() : "";
```
Comment (nornagon) [inline shell/renderer/api/electron_api_web_frame.cc:918]: I don't love reaching into the private Blink APIs here. Further, the `GetWindowFeatures` function is only used in one (relatively obscure) place in Blink, so it seems liable to be removed. Can we do this some other way?

Comment (samuelmaddock) [reply shell/renderer/api/electron_api_web_frame.cc:918]: It's not ideal, the path I used to get the window feature data is the only one I'm aware of.

I believe the code where the window features are saved is a stable area though. See https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/page/create_window.cc;l=352;drc=f51c47ef6c37f12d898997ba00010eb7fba0d090

If the window features were to be removed from the `Page` object, I believe restoring its availability in create_window.cc (linked above) would be straightforward. If we're okay with this bet, I can include a comment above the highlighted code to mention this context for any future upgrade maintainer.

An alternative approach could be to duplicate the `raw_features` string somewhere in blink to then be accessed by our API.

**Comment by samuelmaddock** [comment]
Adding `wip` label until I have time to investigate alternatives which may be a more stable implementation choice.

**Pull request closed without merging**